README.md

```diff
@@ -185,8 +185,10 @@ local on_attach = function(client, bufnr)
         autocmd CursorMoved <buffer> lua vim.lsp.buf.clear_references()
       augroup END
     ]]
+  end
 end
 
+local servers = { "pyright" }
```

Comment:
I am confused by this line.

Prior to seeing this, I thought that no matter my LSP server configuration, I could blindly paste this snippet to map the LSP functionality to sensible (or even standard) nvim keybindings.

Now that I see a reference to a specific LSP server, I wonder if instead I have to edit this snippet to match my specific config?

Reply:
The relevant part of my init.vim currently looks like this:

```vim
lua << EOF
require'lspconfig'.bashls.setup{}
require'lspconfig'.cssls.setup{}
require'lspconfig'.dockerls.setup{}
require'lspconfig'.gopls.setup{}
require'lspconfig'.tsserver.setup{}
require'lspconfig'.yamlls.setup{}
require'lspconfig'.jsonls.setup {
    commands = {
      Format = {
        function()
          vim.lsp.buf.range_formatting({},{0,0},{vim.fn.line("$"),0})
        end
      }
    }
}
require'lspconfig'.pyls.setup{}
EOF

lua << EOF
local nvim_lsp = require('lspconfig')
local on_attach = function(client, bufnr)
  local function buf_set_keymap(...) vim.api.nvim_buf_set_keymap(bufnr, ...) end
  local function buf_set_option(...) vim.api.nvim_buf_set_option(bufnr, ...) end

  buf_set_option('omnifunc', 'v:lua.vim.lsp.omnifunc')

  -- Mappings.
  local opts = { noremap=true, silent=true }
  buf_set_keymap('n', 'gD', '<Cmd>lua vim.lsp.buf.declaration()<CR>', opts)
  buf_set_keymap('n', 'gd', '<Cmd>lua vim.lsp.buf.definition()<CR>', opts)
  buf_set_keymap('n', 'K', '<Cmd>lua vim.lsp.buf.hover()<CR>', opts)
  buf_set_keymap('n', 'gi', '<cmd>lua vim.lsp.buf.implementation()<CR>', opts)
  buf_set_keymap('n', '<C-k>', '<cmd>lua vim.lsp.buf.signature_help()<CR>', opts)
  buf_set_keymap('n', '<space>wa', '<cmd>lua vim.lsp.buf.add_workspace_folder()<CR>', opts)
  buf_set_keymap('n', '<space>wr', '<cmd>lua vim.lsp.buf.remove_workspace_folder()<CR>', opts)
  buf_set_keymap('n', '<space>wl', '<cmd>lua print(vim.inspect(vim.lsp.buf.list_workspace_folders()))<CR>', opts)
  buf_set_keymap('n', '<space>D', '<cmd>lua vim.lsp.buf.type_definition()<CR>', opts)
  buf_set_keymap('n', '<space>rn', '<cmd>lua vim.lsp.buf.rename()<CR>', opts)
  buf_set_keymap('n', 'gr', '<cmd>lua vim.lsp.buf.references()<CR>', opts)
  buf_set_keymap('n', '<space>e', '<cmd>lua vim.lsp.diagnostic.show_line_diagnostics()<CR>', opts)
  buf_set_keymap('n', '[d', '<cmd>lua vim.lsp.diagnostic.goto_prev()<CR>', opts)
  buf_set_keymap('n', ']d', '<cmd>lua vim.lsp.diagnostic.goto_next()<CR>', opts)
  buf_set_keymap('n', '<space>q', '<cmd>lua vim.lsp.diagnostic.set_loclist()<CR>', opts)

  -- Set some keybinds conditional on server capabilities
  if client.resolved_capabilities.document_formatting then
    buf_set_keymap("n", "<space>f", "<cmd>lua vim.lsp.buf.formatting()<CR>", opts)
  elseif client.resolved_capabilities.document_range_formatting then
    buf_set_keymap("n", "<space>f", "<cmd>lua vim.lsp.buf.formatting()<CR>", opts)
  end
  
  -- Set autocommands conditional on server_capabilities
  if client.resolved_capabilities.document_highlight then
    require('lspconfig').util.nvim_multiline_command [[
      :hi LspReferenceRead cterm=bold ctermbg=red guibg=LightYellow
      :hi LspReferenceText cterm=bold ctermbg=red guibg=LightYellow
      :hi LspReferenceWrite cterm=bold ctermbg=red guibg=LightYellow
      augroup lsp_document_highlight
        autocmd!
        autocmd CursorHold <buffer> lua vim.lsp.buf.document_highlight()
        autocmd CursorMoved <buffer> lua vim.lsp.buf.clear_references()
      augroup END
    ]]
end

for _, lsp in ipairs(servers) do
  nvim_lsp[lsp].setup { on_attach = on_attach }
end
EOF
```

If I have to populate the `servers` local manually, then please make it more explicit in the README

Reply:
Previously servers was nil, I just wanted to be more explicit that you should define it as an array. You can put whatever (valid) server names from CONFIG.md in there you want to use. I just chose pyright as an example because it's easy to install, and a large number of people ask about/use python language servers.

Reply:
I think given that I replaced this, it eliminates any suggestion that you *don't* have to call setup manually for each server, I added a loop in the example as most of the time, your keybindings will be consistent across servers, and it's a common pattern to use on_attach to conditionally bind the servers to buffers. In every other example in the readme we manually call setup with a named key for the server. It's possible to setup all of the servers, but it's a bad idea given that we have multiple servers operating on the same filetype (pyls_ms, pyls, pyright, jedi-language-server for example)

Reply:
and in case someone else was wondering how to add multiple items in a LUA array literal: https://www.lua.org/pil/11.1.html

:grin:

Reply:
I'll add 2 for induction's sake lol

Reply:
I suppose that this wouldn't work?

```lua
for _, lsp in ipairs(nvim_lsp) do
  lsp.setup { on_attach = on_attach }
end
```

Reply:
No, you can check the metatable in configs.lua to see why. If you don't call setup, nvim_lsp's namespace doesn't contain the server names.  You need to call setup.py first.

Reply:
It's also not a good idea, as you'll launch 2 rust servers for rust, 4 python servers for python, between 3 and 5 for typescript (I think) amongst other issues.

Reply:
Proposed in https://github.com/neovim/nvim-lspconfig/pull/561

Reply:
Yikes I only saw your comments after refreshing the page.

I assumed that `nvim_lsp` contained all configured servers; I now understand that it contains _all_ servers.

Don't we have a table with the servers the user has required `'lspconfig'` for somewhere? It's strange that I have to first `require'lspconfig'.bashls.setup{}`, and then add `bashls` a few lines below to add keybinding

Reply:
It doesn't contain all servers, it contains the configurations for them. It's not a preset table, because you have the ability to override them via "setup".


>It's strange that I have to first require'lspconfig'.bashls.setup{}, and then add bashls a few lines below to add keybinding

You don't have to `require'lspconfig'.bashls.setup{}` and then add bashls again, you can add on_attach in the function call `require'lspconfig'.bashls.setup{on_attach=on_attach}` Where d

if you print 'lua print(vim.inspect(require('lspconfig').available_servers()))', you'll get a list of which servers have been configured via a "setup" call

Reply:
oh OK, thanks.

So the entire LSP-related section of my init.vim file now looks like:

```vim
lua << EOF
local nvim_lsp = require('lspconfig')
local on_attach = function(client, bufnr)
  local function buf_set_keymap(...) vim.api.nvim_buf_set_keymap(bufnr, ...) end
  local function buf_set_option(...) vim.api.nvim_buf_set_option(bufnr, ...) end

  buf_set_option('omnifunc', 'v:lua.vim.lsp.omnifunc')

  -- Mappings.
  local opts = { noremap=true, silent=true }
  buf_set_keymap('n', 'gD', '<Cmd>lua vim.lsp.buf.declaration()<CR>', opts)
  buf_set_keymap('n', 'gd', '<Cmd>lua vim.lsp.buf.definition()<CR>', opts)
  buf_set_keymap('n', 'K', '<Cmd>lua vim.lsp.buf.hover()<CR>', opts)
  buf_set_keymap('n', 'gi', '<cmd>lua vim.lsp.buf.implementation()<CR>', opts)
  buf_set_keymap('n', '<C-k>', '<cmd>lua vim.lsp.buf.signature_help()<CR>', opts)
  buf_set_keymap('n', '<space>wa', '<cmd>lua vim.lsp.buf.add_workspace_folder()<CR>', opts)
  buf_set_keymap('n', '<space>wr', '<cmd>lua vim.lsp.buf.remove_workspace_folder()<CR>', opts)
  buf_set_keymap('n', '<space>wl', '<cmd>lua print(vim.inspect(vim.lsp.buf.list_workspace_folders()))<CR>', opts)
  buf_set_keymap('n', '<space>D', '<cmd>lua vim.lsp.buf.type_definition()<CR>', opts)
  buf_set_keymap('n', '<space>rn', '<cmd>lua vim.lsp.buf.rename()<CR>', opts)
  buf_set_keymap('n', 'gr', '<cmd>lua vim.lsp.buf.references()<CR>', opts)
  buf_set_keymap('n', '<space>e', '<cmd>lua vim.lsp.diagnostic.show_line_diagnostics()<CR>', opts)
  buf_set_keymap('n', '[d', '<cmd>lua vim.lsp.diagnostic.goto_prev()<CR>', opts)
  buf_set_keymap('n', ']d', '<cmd>lua vim.lsp.diagnostic.goto_next()<CR>', opts)
  buf_set_keymap('n', '<space>q', '<cmd>lua vim.lsp.diagnostic.set_loclist()<CR>', opts)

  -- Set some keybinds conditional on server capabilities
  if client.resolved_capabilities.document_formatting then
    buf_set_keymap("n", "<space>f", "<cmd>lua vim.lsp.buf.formatting()<CR>", opts)
  elseif client.resolved_capabilities.document_range_formatting then
    buf_set_keymap("n", "<space>f", "<cmd>lua vim.lsp.buf.formatting()<CR>", opts)
  end
  
  -- Set autocommands conditional on server_capabilities
  if client.resolved_capabilities.document_highlight then
    require('lspconfig').util.nvim_multiline_command [[
      :hi LspReferenceRead cterm=bold ctermbg=red guibg=LightYellow
      :hi LspReferenceText cterm=bold ctermbg=red guibg=LightYellow
      :hi LspReferenceWrite cterm=bold ctermbg=red guibg=LightYellow
      augroup lsp_document_highlight
        autocmd!
        autocmd CursorHold <buffer> lua vim.lsp.buf.document_highlight()
        autocmd CursorMoved <buffer> lua vim.lsp.buf.clear_references()
      augroup END
    ]]
  end
end

require'lspconfig'.bashls.setup{on_attach=on_attach}
require'lspconfig'.cssls.setup{on_attach=on_attach}
require'lspconfig'.dockerls.setup{on_attach=on_attach}
require'lspconfig'.gopls.setup{on_attach=on_attach}
require'lspconfig'.tsserver.setup{on_attach=on_attach}
require'lspconfig'.yamlls.setup{on_attach=on_attach}
require'lspconfig'.jsonls.setup {
    commands = {
      Format = {
        function()
          vim.lsp.buf.range_formatting({},{0,0},{vim.fn.line("$"),0})
        end
      }
    },
    on_attach=on_attach
}
require'lspconfig'.pyls.setup{on_attach=on_attach}
EOF
```

and it seems to work.

Would you accept a PR where I move this `on_attach` function into the plugin codebase?

Reply:
on_attach *is* in the plugin codebase. We're just overriding it with a custom function which defines the buffer specific keymaps.  

What you just did above is a more verbose version of what's already in the readme. You could have just done it in a loop like I did.

The point of the custom keymaps is to show how on_attach can be leveraged in one particular way. We don't (and won't) provide default keybindings within the code, as this is very personal. I also personally override my on attach for different servers, so I would find it annoying that the base on_attach was opinionated.

I would not accept a PR changing on attach to map keybindings by default.

Reply:
> What you just did above is a more verbose version of what's already in the readme. You could have just done it in a loop like I did.

OK now I understand that the `require` work is already coded in the `on_attach` function in the snippet, and I don't have to FIRST require and THEN attach the keybindings. Hence I have removed the `require` part and just let in the `on_attach` definition with the keybindings, and the loop with the names of the servers.

Thanks a lot!

Reply:
Just to be clear (for your future lua endeavours) lua cache's module loading, so the penalty is low for multiple requires, it's more a sylistic concern. On_attach does not contain a `require`, we call that before the function definition in the parent scope.